tsdb/shard.go

```diff
@@ -1302,6 +1306,143 @@ func (a Shards) FieldKeysByPredicate(expr influxql.Expr) (map[string][]string, e
+			if v.onlyAnds {
+				v.head = n
+			}
+			v.onlyAnds = false
+		}
+
+		switch n.Op {
+		case influxql.AND:
+			// If only AND ops have been encountered so far, move the head of this
+			// subtree along and continue.
+			if v.onlyAnds {
+				v.head = n
+				return v
+			}
+
+			// Otherwise, encountering this AND after something other than an AND
+			// invalidates this subtree.
+			v.invalidHeads[v.head] = struct{}{}
+			return v
+		case influxql.EQ:
+			// For EQ ops, the only valid configuration is the form of
+			// "_measurement = measurementName", so check for that and add the
+			// resulting name to the list of names for this subtree.
+			if meas, ok := measurementNameFromEqual(n, v.measurementKey); ok {
+				v.measurementNodeHeads[v.head] = append(v.measurementNodeHeads[v.head], meas)
+				return v
+			}
+
+			// If it wasn't a measurement, this tree is invalid.
+			v.invalidHeads[v.head] = struct{}{}
+			return v
+		case influxql.OR:
+			// OR ops do not invalidate the subtree.
+			return v
+		default:
+			// Any other operation means this subtree is invalid.
+			v.invalidHeads[v.head] = struct{}{}
+
+			// If a measurement name is being operated on with a non-equality
+			// operator, add the measurement name to the subtree. This must be done to
+			// indicate that there is a measurement in an invalid subtree.
+			if ref, ok := n.LHS.(*influxql.VarRef); ok {
+				if ref.Val == v.measurementKey {
+					v.measurementNodeHeads[v.head] = append(v.measurementNodeHeads[v.head], ref.Val)
+				}
+			}
+
+			return v
+		}
+	case *influxql.VarRef, *influxql.StringLiteral, *influxql.ParenExpr:
+		// VarRefs and StringLiterals will have been evaluated in the EQ case.
+		// Parens also do not invalidate the subtree and can be traversed through.
+		return v
+	}
+
+	// If the function did not return in the type switch, that means that this is
+	// some other kind of node, which invalidates the subtree.
+	v.onlyAnds = false
+	v.invalidHeads[v.head] = struct{}{}
+
+	return v
+}
+
+func measurementOptimization(expr influxql.Expr, key string) ([][]byte, bool) {
+	measurementNodeHeads := make(map[*influxql.BinaryExpr][]string)
+	invalidHeads := make(map[*influxql.BinaryExpr]struct{})
+	dummyHead := &influxql.BinaryExpr{}
+
+	v := measurementEvaluator{
+		measurementNodeHeads: measurementNodeHeads,
+		invalidHeads:         invalidHeads,
+		onlyAnds:             true,
+		head:                 dummyHead,
+		measurementKey:       key,
+	}
+
+	influxql.Walk(v, expr)
+
+	// There must be measurements in exactly one subtree for this optimization to
+	// work. Note: It may be possible to have measurements in multiple subtrees,
+	// as long as there are no measurements in invalid subtrees, by determining an
+	// intersection of the measurement names across all valid subtrees - this is
+	// not currently implemented.
+	if len(measurementNodeHeads) != 1 {
+		return nil, false
+	}
+
+	names := []string{}
+	for h, ns := range measurementNodeHeads {
```

Comment:
At this point we are guaranteed `len(measurementNodeHeads) == 1` - should just check for invalidHeads.